packages/utils/src/lock.ts

```diff
@@ -0,0 +1,54 @@
+import { EventEmitter } from 'events'
+
+enum LockEvent {
+  Unlock = 'unlock',
+}
+
+// Lock which can be used to ensure mutual exclusion in concurrent code.
+export class Lock {
+  private locked: boolean = false
+  private emitter: EventEmitter
+
+  constructor() {
+    this.emitter = new EventEmitter()
+  }
+
+  // Attempt to aquire the lock without blocking.
+  //
+  // @returns {boolean} True if the lock was aquired.
+  tryAcquire(): boolean {
+    if (!this.locked) {
+      this.locked = true
+      return true
+    }
+    return false
+  }
+
+  // Aquire the lock, blocking until the lock is available.
+  acquire(): Promise<void> {
+    return new Promise((resolve) => {
+      // Attempt to grab the lock without waiting.
+      if (this.tryAcquire()) {
+        resolve()
+        return
+      }
+
+      // Wait for an event emitted when releasing the lock.
+      const callback = () => {
+        if (this.tryAcquire()) {
+          this.emitter.removeListener(LockEvent.Unlock, callback)
+          resolve()
+        }
+      }
+      this.emitter.on(LockEvent.Unlock, callback)
```

Comment:
curious -- what happens if you have multiple requests calling `emitter.on`?  

if an unlock happens and all their callbacks are triggered, only one will successfully acquire, and it will remove the listener (hopefully not for all pending requests...).

Reply:
Because the callback is created in the function call, the function (and it's closure) are identity unique, so only the callback for the one that will acquire the lock will be removed. (If this were not the case, the race in the unit test would deadlock.)

Reply:
i see -- so it just removed that callback rather than full listener.  makes sense 👍

Reply:
i see -- so it just removed that callback rather than full listener. makes sense 👍